components/raftstore-v2/src/operation/command/admin/merge/prepare.rs

```diff
@@ -0,0 +1,494 @@
+                self.logger.list()
+            ));
+        }
+        self.validate_prepare_merge_command(
+            store_ctx,
+            req.get_admin_request().get_prepare_merge(),
+        )?;
+        let pre_propose = if let Some(r) = self.already_checked_pessimistic_locks()? {
+            r
+        } else {
+            let r = self.check_logs_before_prepare_merge(store_ctx)?;
+            self.check_pessimistic_locks(r, &mut req)?
+        };
+        req.mut_admin_request()
+            .mut_prepare_merge()
+            .set_min_index(pre_propose.min_matched + 1);
+        self.propose_locks_before_prepare_merge(store_ctx, pre_propose.lock_size_limit)
+            .and_then(|_| {
+                let mut proposal_ctx = ProposalContext::empty();
+                proposal_ctx.insert(ProposalContext::PREPARE_MERGE);
+                let data = req.write_to_bytes().unwrap();
+                self.propose_with_ctx(store_ctx, data, proposal_ctx.to_vec())
+            })
+            .inspect_err(|_| {
+                // Match v1::post_propose_fail.
+                // If we just failed to propose PrepareMerge, the pessimistic locks status
+                // may become MergingRegion incorrectly. So, we have to revert it here.
+                // But we have to rule out the case when the region has successfully
+                // proposed PrepareMerge or has been in merging, which is decided by
+                // the boolean expression below.
+                if !self.proposal_control().is_merging() {
+                    let mut pessimistic_locks = self.txn_context().ext().pessimistic_locks.write();
+                    if pessimistic_locks.status == LocksStatus::MergingRegion {
+                        pessimistic_locks.status = LocksStatus::Normal;
+                    }
+                }
+                self.free_merge_context();
+            })
+    }
+
+    /// Match v1::check_merge_proposal.
+    /// - Target region epoch as requested is identical with the local version.
+    /// - Target region is a sibling to the source region.
+    /// - Peers of both source and target region are aligned, i.e. located on
+    ///   the same set of stores.
+    fn validate_prepare_merge_command<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        req: &PrepareMergeRequest,
+    ) -> Result<()> {
+        // Just for simplicity, do not start region merge while in joint state
+        if self.in_joint_state() {
+            return Err(box_err!(
+                "{} region in joint state, can not propose merge command, command: {:?}",
+                SlogFormat(&self.logger),
+                req
+            ));
+        }
+        let region = self.region();
+        let target_region = req.get_target();
+        {
+            let store_meta = store_ctx.store_meta.lock().unwrap();
+            match store_meta.readers.get(&target_region.get_id()) {
```

Comment:
Why not store_meta.regions.get...

Reply:
Good, that field doesn't exist when I wrote the code.